### ISSUE TITLE
E2E failure due to openstack cli python deps?

### ISSUE THREAD

**Issue by mdbooth** [kubernetes-sigs/cluster-api-provider-openstack#1351]
/kind bug

**What steps did you take and what happened:**
This job failed: https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-openstack/1350/pull-cluster-api-provider-openstack-e2e-test/1576886172808908800/build-log.txt

The relevant trace from the build log looks like:

```
++ openstack versions show
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/stevedore/_cache.py", line 159, in _get_data_for_path
    with open(filename, 'r') as f:
FileNotFoundError: [Errno 2] No such file or directory: '/root/.cache/python-entrypoints/aa1e52575b0b33cd4df0aed2af56d2b5dfa248c667d0055ed45042f5ce9d68b9'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/bin/openstack", line 5, in <module>
    from openstackclient.shell import main
  File "/usr/local/lib/python3.7/dist-packages/openstackclient/shell.py", line 26, in <module>
    from openstackclient.common import clientmanager
  File "/usr/local/lib/python3.7/dist-packages/openstackclient/common/clientmanager.py", line 198, in <module>
    'openstack.cli.base',
  File "/usr/local/lib/python3.7/dist-packages/openstackclient/common/clientmanager.py", line 149, in get_plugin_modules
    mgr = stevedore.ExtensionManager(group)
  File "/usr/local/lib/python3.7/dist-packages/stevedore/extension.py", line 136, in __init__
    verify_requirements)
  File "/usr/local/lib/python3.7/dist-packages/stevedore/extension.py", line 218, in _load_plugins
    for ep in self.list_entry_points():
  File "/usr/local/lib/python3.7/dist-packages/stevedore/extension.py", line 207, in list_entry_points
    eps = list(_cache.get_group_all(self.namespace))
  File "/usr/local/lib/python3.7/dist-packages/stevedore/_cache.py", line 179, in get_group_all
    data = self._get_data_for_path(path)
  File "/usr/local/lib/python3.7/dist-packages/stevedore/_cache.py", line 162, in _get_data_for_path
    data = _build_cacheable_data(path)
  File "/usr/local/lib/python3.7/dist-packages/stevedore/_cache.py", line 110, in _build_cacheable_data
    for name, group_data in real_groups.items():
AttributeError: 'EntryPoints' object has no attribute 'items'
+ echo 'failed 1 times: openstack versions show'
failed 1 times: openstack versions show
```

This looks like an issue in the installation of openstack cli.

**Comment by mdbooth** [comment]
I can easily reproduce the issue locally:

```
> podman run --rm -ti debian:buster
root@aa9d05f761bc:/# apt-get update
...snip
root@aa9d05f761bc:/# apt-get install -y python3-virtualenv
...snip
root@aa9d05f761bc:/# python3 -m virtualenv -p $(which python3) /tmp/openstack
...snip
root@aa9d05f761bc:/# source /tmp/openstack/bin/activate
(openstack) root@aa9d05f761bc:/# pip3 install python-cinderclient python-glanceclient python-keystoneclient python-neutronclient python-novaclient python-openstackclient python-octaviaclient
(openstack) root@aa9d05f761bc:/# openstack versions show
Traceback (most recent call last):
  File "/tmp/openstack/lib/python3.7/site-packages/stevedore/_cache.py", line 159, in _get_data_for_path
    with open(filename, 'r') as f:
FileNotFoundError: [Errno 2] No such file or directory: '/root/.cache/python-entrypoints/d95bf8bde8c40006489f65cf353c1ca2fbbdccfd670b2606147bd25021ab52a0'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/openstack/bin/openstack", line 5, in <module>
    from openstackclient.shell import main
  File "/tmp/openstack/lib/python3.7/site-packages/openstackclient/shell.py", line 26, in <module>
    from openstackclient.common import clientmanager
  File "/tmp/openstack/lib/python3.7/site-packages/openstackclient/common/clientmanager.py", line 198, in <module>
    'openstack.cli.base',
  File "/tmp/openstack/lib/python3.7/site-packages/openstackclient/common/clientmanager.py", line 149, in get_plugin_modules
    mgr = stevedore.ExtensionManager(group)
  File "/tmp/openstack/lib/python3.7/site-packages/stevedore/extension.py", line 136, in __init__
    verify_requirements)
  File "/tmp/openstack/lib/python3.7/site-packages/stevedore/extension.py", line 218, in _load_plugins
    for ep in self.list_entry_points():
  File "/tmp/openstack/lib/python3.7/site-packages/stevedore/extension.py", line 207, in list_entry_points
    eps = list(_cache.get_group_all(self.namespace))
  File "/tmp/openstack/lib/python3.7/site-packages/stevedore/_cache.py", line 179, in get_group_all
    data = self._get_data_for_path(path)
  File "/tmp/openstack/lib/python3.7/site-packages/stevedore/_cache.py", line 162, in _get_data_for_path
    data = _build_cacheable_data(path)
  File "/tmp/openstack/lib/python3.7/site-packages/stevedore/_cache.py", line 110, in _build_cacheable_data
    for name, group_data in real_groups.items():
AttributeError: 'EntryPoints' object has no attribute 'items'
```

**Comment by stephenfin** [comment]
This is Python issue of sorts. I suspect if you ran `pip freeze` in that container you would see the following:

```
importlib-metadata==5.0.0
```

`importlib-metadata` is a backport of the `importlib.metadata` library from stdlib. They had deprecated the dictionary-style interface for `EntryPoint` and [recently removed this entirely as part of 5.0.0](https://github.com/python/importlib_metadata/commit/dde2b9de2973ce1c6fa9ba21dfe81069b0baa77b). `stevedore` needs to adapt here and hasn't had to do this already since we don't support Python 3.7 anymore and the stdlib version (which has been present since Python 3.8) is good enough for us.

There are a couple of ways around this. The first and most sensible is to bump the container image to something that supports Python 3.8 or newer. The second is to use OpenStack's requirement constraints system.

```bash
$ pip install -r https://releases.openstack.org/constraints/upper/yoga \
    python-cinderclient \
    python-glanceclient \
    python-keystoneclient \
    python-neutronclient \
    python-novaclient \
    python-openstackclient \
    python-octaviaclient
```

(we use the yoga release since that's the last one to support Python 3.7 (albeit unofficially - it technically supported 3.6 and 3.8).

The last is to simply pin the version of `importlib-metadata` temporarily until we fix `stevedore` (and ideally backport the fix to `stable/yoga`).

**Comment by mdbooth** [comment]
I started looking into bumping the test runner image. IIUC (a serious consideration):

* The image used is [kubekins-e2e, defined in the presubmit job](https://github.com/kubernetes/test-infra/blob/63d1bec450c4f5f657f0bfecd1a6191229bfe0f1/config/jobs/kubernetes-sigs/cluster-api-provider-openstack/cluster-api-provider-openstack-presubmits.yaml#L60).
* [kubekins-e2e is defined here](https://github.com/kubernetes/test-infra/tree/master/images/kubekins-e2e).
* kubekins-e2e is based on [the bootstrap image](https://github.com/kubernetes/test-infra/blob/63d1bec450c4f5f657f0bfecd1a6191229bfe0f1/images/kubekins-e2e/Dockerfile#L20).
* Assuming I've found the right one, [the bootstrap image is also defined in test-infra](https://github.com/kubernetes/test-infra/tree/master/images/bootstrap).
* According to the README of the bootstrap image, it's deprecated.

IOW, changing that looks like a potentially large job for somebody with a lot of experience in this area. We should likely coordinate this with other projects.

**Comment by mdbooth** [comment]
@stephenfin I've implemented something based on your suggestion in #1352. Lets see if it passes. Would appreciate your review.
